### DISCUSSION TITLE
Using Stream to monitor /dev/ttyS0 ??

### DISCUSSION THREAD

**Discussion by michaelphipps** [reactphp/reactphp#483] [Q&A] [answered]
I've got a  Sim hat for my Raspberry Pi.  I've been using a modified version of https://github.com/Xowap/PHP-Serial to read and write to the serial port to send AT Commands and receive responses.  

Is this the type of thing I can use react/stream for?

**Comment by michaelphipps** [comment, marked as answer]
OMG!  Is it REALLY this easy?

```php
<?php
use React\Stream\DuplexResourceStream;

require "vendor/autoload.php";

$resource = fopen("/dev/ttyS0", "w+");

$stream = new DuplexResourceStream($resource);

$stream->on('data', function ($chunk) {
    echo $chunk;
});
```

**Reply by SimonFrings** [reply]
@michaelphipps Seems like a nice solution, if it works properly then everything is fine :+1:

It seems to me like you answered your own question, thanks for sharing your experience. If you have any more questions let me know.

**Reply by michaelphipps** [reply]
For anyone that stumbles across this, for my particular use (talking to a SIM7600X connected to a raspberry pi) I needed to change the characteristics of the terminal using `stty` before setting up the resource.

```
exec("stty -F /dev/ttyS0 115200 -parenb cs8 -cstopb clocal -crtscts -ixon -ixoff -echo");
```
what this line does:
* sets the baud rate to 115200 [115200]
* sets the parity to none [-parenb]
* sets the character length to 8 [cs8]
* sets the stop bits to 1 [-cstopb]
* sets the flow control to none [clocal -crtscts -ixon -ixoff]
* turns of echo  so that any data I send is not read by `$stream->on('data', ...)` [-echo]

Keep in mind data is not received line by line from the terminal.  You need to reassemble the chunks of data you receive into something usable yourself.
